File: Data Structures & Algorithms/search-2d-matrix/submission-4.py
class Solution:
    def searchMatrix(self, matrix: List[List[int]], target: int) -> bool:
        m, n = len(matrix), len(matrix[0])
        s, e = 0, m * n - 1
        while s <= e:
            i = e - s // 2
            x, y = i // n, i % n
            curr = matrix[x][y]
            if curr == target:
                return True
            if curr < target:
                s = i+1
            else:
                e = i-1
        return False

File: Data Structures & Algorithms/car-fleet/submission-4.ts
class Solution {
    /**
     * @param {number} target
     * @param {number[]} position
     * @param {number[]} speed
     * @return {number}
     */
    carFleet(target: number, position: number[], speed: number[]): number {
        const combined = position.map((p, idx) =>  [p, speed[idx]]).sort(([a], [b]) => b - a)
        const fleets = combined.reduce((fleets, [p, s]) => {
            const count = (target-p) / s
            if (!fleets.length) {
                fleets.push(count)
            } else {
                const top = fleets[fleets.length - 1]
                if (count > top) {
                    fleets.push(count)
                }

            }
            return fleets
        }, [])

        return fleets.length

    }
}
